contracts/SimpleGAMultiSig.aes

```diff
@@ -0,0 +1,154 @@
+        require(tx.ga_tx_hash == ga_tx_hash, "ERROR_UNEQUAL_HASHES")
+        Chain.event(TxAuthorized(ga_tx_hash))
+        put(state{ current_tx = None })
+        true
+
+  // propose a new tx valid for the given ttl
+  stateful entrypoint propose(ga_tx_hash : hash, ttl : Chain.ttl, sig : signature) =
+    require_valid_signer_and_sig(Call.caller, ga_tx_hash, sig)
+    switch(state.current_tx)
+      None => set_current_tx(ga_tx_hash, ttl)
+      Some(tx) =>
+        require(is_expired(tx), "ERROR_A_TX_IS_ALREADY_PROPOSED")
+        set_current_tx(ga_tx_hash, ttl)
+
+  // signer confirms tx
+  stateful entrypoint confirm(sig : signature) =
+    let tx = Option.force_msg(state.current_tx, "ERROR_NOTHING_TO_CONFIRM")
+    require_valid_signer_and_sig(Call.caller, tx.ga_tx_hash, sig)
+    require(!Set.member(Call.caller, tx.confirmed_by), "ERROR_ALREADY_CONFIRMED")
+    require(!is_expired(tx), "ERROR_TX_ALREADY_EXPIRED")
+    let updated_tx = tx{ confirmed_by = Set.insert(Call.caller, tx.confirmed_by) }
+    put(state{ current_tx = Some(updated_tx) })
+    Chain.event(TxConfirmed(tx.ga_tx_hash, Call.caller))
+    if(has_consensus(updated_tx))
+      Chain.event(TxConsensusReached(tx.ga_tx_hash))
+
+  // revoke the current tx and clean state
+  stateful entrypoint revoke(sig : signature) =
+    let tx = Option.force_msg(state.current_tx, "ERROR_NOTHING_TO_REVOKE")
+    require_valid_signer_and_sig(Call.caller, tx.ga_tx_hash, sig)
+    put(state{ current_tx = None })
+    Chain.event(TxRevoked(tx.ga_tx_hash, Call.caller))
+
+  // return all signers
+  entrypoint get_signers() : list(address) =
+  	Set.to_list(state.signers)
+
+  // return the consensus info
+  entrypoint get_consensus_info() : consensus_info =
+    switch(state.current_tx)
+      None => { ga_tx_hash = None,
+                confirmations_required = state.confirmations_required,
+                confirmed_by = [],
+                has_consensus = false,
+                expired = false }
+      Some(tx) => { ga_tx_hash = Some(tx.ga_tx_hash),
+                    confirmations_required = state.confirmations_required,
+                    confirmed_by = Set.to_list(tx.confirmed_by),
+                    has_consensus = has_consensus(tx),
+                    expired = is_expired(tx) }
+
+  // validate if enough signers have confirmed
+  function has_consensus(tx: transaction) : bool =
+    Set.size(tx.confirmed_by) >= state.confirmations_required
+
+  // require that given address is within the list of signers
+  // AND
+  // that the signature of the proposed ga tx was made using the private key of the given address
+  function require_valid_signer_and_sig(addr : address, ga_tx_hash_to_verify : hash, sig : signature) =
+    require(Set.member(addr, state.signers) && Crypto.verify_sig(ga_tx_hash_to_verify, addr, sig), "ERROR_NOT_AUTHORIZED")
+
+  // check if current_tx is already expired
+  function is_expired(tx: transaction) : bool =
```

Comment:
most times (stateful) used in negated form, maybe make that the default

Reply:
you mean we should change to `is_not_expired` ?

Reply:
yeah, or `not_expired` but just my feeling not necessarily needed, its easier to read than `!is_expired` as you could easily missread that or miss an `!` in logic